Check newly imported coding system releases for compatibility with existing codelist versions

rebkwok opened the pull request:
Adds a `compatible_releases` field to CodelistVersion, so every CodelistVersion now knows:
1) which coding system release it was created with and
2) which subsequent coding system releases it is compatible with

Compatibility is checked after importing a new release, by looking for all CodelistVersions that are compatible/created with the _previous_ release (determined by the `valid_from` date).  If a version wasn't compatible with the previous release, we assume it also isn't compatible with this release.  This should mean that the number of versions that we need to check after each import is manageable (for the first new import, where every existing codelist version is marked as "unknown", we'll need to check all of them; for snomed, this is ~3000 versions, and takes a few minutes).

A CodelistVersion is deemed to be compatible with a new release if:
- It has searches, and rerunning the searches with the new release returns identical results OR
- It has a cached hierarchy (this applies to all coding systems that we can import new releases for, except dm+d), and a hierarchy created from its codes with the new release is identical 

The hierarchy check is more restrictive - e.g. there could be a new parent code in the hierarchy that isn't returned by a search, so we do the search check first, and the hierarchy check only on codelists that don't have searches.

On the version and builder pages, the codelist version details also list compatible releases (in an expandable section, since it could be quite a long list for some versions, especially when dm+d ones - with weekly releases - have hierarchies/searches):

![Screenshot from 2022-11-29 16-31-20](https://user-images.githubusercontent.com/6770950/204590510-b3a8ec17-be3e-42b1-978d-c4efe402100c.png)

Fixes #1441

inglesp commented on coding_systems/base/import_data_utils.py:
```diff
@@ -136,3 +149,118 @@ def setup_new_import_database(self):
+        CodingSystemRelease.objects.filter(coding_system=coding_system_id)
+        .exclude(id=new_coding_system.release.id)
+        .latest("valid_from")
+    )
```
I'm quite paranoid...  I feel it would be helpful to assert here that `previous_release.valid_from` is earlier than new_coding_system.release.valid_from`.

rebkwok replied:
hmmm...it should be.  Let me check if I've forgotten to push something

rebkwok replied:
I've added that now - I had updated it so it filtered by valid_from < new_coding_system.release.valid_from, but that seems to have gone missing in a rebase somewhere

inglesp commented on coding_systems/base/import_data_utils.py:
```diff
@@ -136,3 +149,118 @@ def setup_new_import_database(self):
+        versions_to_check.filter(
+            codelist__coding_system_id=coding_system_id, searches__isnull=False
+        )
+    )
```
I don't think this is enough: you can have a codelist version that was created by uploading a CSV, and where a user has done some searches, but where these searches don't cover all of the codes in the codelist.

So should we be only checking the hierarchies?

rebkwok replied:
So should we only check by searches for codelist versions where the searches cover all codes?

inglesp commented on coding_systems/base/import_data_utils.py:
```diff
@@ -136,3 +149,118 @@ def setup_new_import_database(self):
+    return compatible_count, number_of_versions
+
+
+def _check_version_by_search(coding_system, version):
```
This needs a docstring explaining why the searches being the same implies compatibility.  (Although actually I think it doesn't -- see above.)

inglesp commented on coding_systems/base/import_data_utils.py:
```diff
@@ -136,3 +149,118 @@ def setup_new_import_database(self):
+    return True
+
+
+def _check_version_by_hierarchy(coding_system, version):
```
This needs a thorough docstring and some really explicit tests.  My nervousness here is that we're turning some functionality that was designed for to support performance into something that we're depending on for data correctness.

New tests should demonstrate the ways that a full coding system hierarchy can change, and the ways that this changes whether or not a codelist version's hierarchy (a `Hierarchy` instance) needs to change.  By tweaking `Hierarchy.from_codes` so that it calls through to a new class method that doesn't know anything about coding systems, you should be able to test this just by calling this new class method.  I'd be happy to pair on this if you like.

rebkwok commented:
@inglesp This PR is updated and I think ready for another review; I've added commits with new changes rather than rewriting the existing ones; the last 5 are new since you reviewed last.

As discussed, the compatibility check is now by both hierarchies and searches.  A codelist version is only compatible with a new release if the hierarchy built with the new and old release are identical AND the search results with the new and old release are identical.

I added tests for `Hierarchy.from_codes`, using a mock coding system which is essentially a larger hierarchy, which can be used to build the `Hierarchy` from code(s), which is a subset of it, and also tests to verify changes to the coding system that do or don't result in changes to the hierarchy.

I also realised that we only want to check and update compatibility for codelist versions that are under-review or published.  Codes for those won't change, whereas for drafts, they could change, and we don't want to have to re-check compatibility every time a draft is updated. So the compatibility check on importing new data only looks at under-review and published versions, and when a draft is saved for review, it's checked against any newer releases of the coding system at that point.

inglesp reviewed (approved):
Great stuff, thanks.  I want to rewrite all the existing tests with the new mock coding system!

rebkwok commented on coding_systems/base/import_data_utils.py:
```diff
@@ -180,57 +200,95 @@ def update_codelist_version_compatibility(coding_system_id, new_database_alias):
+    # For any codelist with the "unknown" coding system release, we fall back to using
+    # its cached hierarchy.
+    # For any codelist created with a known coding system release, we can re-create its hierarchy
+    # from the original coding system release and use that for comparison with the new release.
```
It is, of course, much slower, so I've changed it back to always using the cache for the original one
